scripts/api.py

```diff
@@ -77,8 +79,29 @@ async def detect(
         processor_module = cached_cn_preprocessors[controlnet_module]
 
         for input_image in controlnet_input_images:
+            
             img = external_code.to_base64_nparray(input_image)
-            results.append(processor_module(img, res=controlnet_processor_res, thr_a=controlnet_threshold_a, thr_b=controlnet_threshold_b)[0])
+
+            res = controlnet_processor_res
+
+            if controlnet_pixel_perfect:
+
+                h, w = img.shape[:2]
+
+                h = (h // 8) * 8
```

Comment:
The casting is not pixel perfect specific in `process`. 

I believe this casting does not appear in `run_annotator` code in uigroup. 
@lllyasviel Is this logic necessary in `run_annotaor` as well?

Reply:
BTW, I borrowed this logic from existing code https://github.com/Mikubill/sd-webui-controlnet/blob/09cb9a32d1051aa827f1bb092cf17fcbf996ed7f/scripts/controlnet.py#L1204

Reply:
I added a couple of tests as well.

Reply:
> BTW, I borrowed this logic from existing code
> 
> https://github.com/Mikubill/sd-webui-controlnet/blob/09cb9a32d1051aa827f1bb092cf17fcbf996ed7f/scripts/controlnet.py#L1204

The existing code is outside `if unit.pixel_perfect:` check.